### PR TITLE
JAMES-3982 Upgrade MinIO image tag for tests

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/ReadSaveBlobStoreDAOContract.java [hunk 1]
```diff
@@ -236,7 +236,7 @@ default void saveBytesShouldBeIdempotent(String description, byte[] bytes) {
     @MethodSource("blobs")
     default void saveByteSourceShouldBeIdempotent(String description, byte[] bytes) {
         BlobStoreDAO store = testee();
-        Mono.from(store.save(TEST_BUCKET_NAME, TEST_BLOB_ID, ByteSource.wrap(bytes))).block();
+        Mono.from(store.save(TEST_BUCKET_NAME, TEST_BLOB_ID, TWELVE_MEGABYTES)).block();
```
Comment (Arsnael) [inline server/blob/blob-api/src/test/java/org/apache/james/blob/api/ReadSaveBlobStoreDAOContract.java:239]: Not sure if it's acceptable or not....

It seems for some reason first save fails if it's with the ByteSource or InputStream with a big blob, but not with the direct bytes. Also if before the two ByteSource big blob saves you do a save on a smaller blob, then it's green.

It's rather odd and I don't really understand what's going on there, if it's related to MinIO warming up or bucket data init... Will try to dig more but... looks rather weird to me.

If someone else with more object storage or MinIO knowledge wants to have a look as well, feel free :)

**Comment by Arsnael** [comment]
Obviously still unstable on the CI... Honestly I don't get what's going on. I guess might need to look on MinIO side... -_-

**Comment by vttranlina** [comment]
After the commit "Allow retries for S3 save" https://github.com/apache/james-project/pull/2025/commits/a0786904da56ff21e4679e30573850b031023221
hope that it is more stable
I tried to repeat test on local, and it passed.

I also test with the newer tag "RELEASE.2024-02-17T01-15-57Z"
But it always failed (throws other error, `Unable to execute HTTP request: The connection was closed during the request.`), after debugging => I think keeping the current tag is simple

**Comment by chibenwa** [comment]
That's... a bitter sweet conclusion...

**Comment by quantranhong1999** [comment]
Maybe we should have a look at this again as we are going to use MinIO.

**Comment by Arsnael** [comment]
@quantranhong1999 I did a few weeks ago and the issue was still the same. But have a try at it if you want :)

**Comment by vttranlina** [comment]
Recently, I had the chance to investigate this issue again, and I have catched the cause.

Starting from Minio `RELEASE.2024-01-31T20-20-33Z`. When we perform a PUT Object request but the bucket does not exist.

The problem is that, for large objects (e.g., 12MB in our test), minio returns a response with a header `Connection: close`, which is absent for smaller objects.
In both object scenarios, the response body is the same `HTTP 404, <Code>NoSuchBucket</Code>`

This header causes AWS SDK S3 to throw an `SdkClientException` (message: "Unable to execute HTTP request: The connection was closed during the request"), which we haven't handled -> test case failure.

Without the `Connection: close` header, the library throws a `NoSuchBucketException`, which we handle by creating the bucket (via `createBucketOnRetry`) -> the putObject then succeeds.


This is why Rene mentioned: "Also if before the two ByteSource big blob saves you do a save on a smaller blob, then it's green." 
I agree this won't happen in production since we create the bucket upfront.

I'm not sure whether this is a bug or a feature of S3 Minio (I’m asking them on their Slack channel and awaiting response).

___________________

Here are some of my thoughts around this issue:

1. Modify the current logic to handle both `NoSuchBucketException` and `SdkClientException`

// this approach doesn’t sound convincing (create bucket when `SdkClientException`)

2. Override the test class S3MinioTest, create the bucket before running the logic.

3. Use S3 multipart upload api for large objects. https://www.baeldung.com/aws-s3-multipart-upload 

4. Experiment with using `S3CrtAsyncClient` instead of `S3AsyncClient`. I quickly researched “crt,” which is a client developed by AWS, and based on the promotion, it sounds very promising for handling large object uploads.

Regarding points (3) and (4), I'm unsure if these changes are worth it, considering our object size limit is 20-30MB.

**Comment by chibenwa** [comment]
Thanks a lot @vttranlina for the diagnostic!

Ok the error looks benign to me.

> 1.  Modify the current logic to handle both NoSuchBucketException and SdkClientException

-1 to me, we would be interpreting a non explicit error, and potentially risk to do a mistake. Bad idea.

> 3. Use S3 multipart upload api for large objects. https://www.baeldung.com/aws-s3-multipart-upload

IMO we "could" experiment with this though I would like expect performance benchmark backing the benefits of multipart uploads.

Though we shall notice that such a fix is mostly accidental and may yield side effects...

Maybe a lot of code changes for a not-that-common situation, though the idea is nice?

> Experiment with using S3CrtAsyncClient instead of S3AsyncClient. I quickly researched “crt,” which is a client developed by AWS, and based on the promotion, it sounds very promising for handling large object uploads.

We do not have that of a big workload regarding large objects to be fair.

Changing of driver is a big move, so overall -1 with me.

> 2. Override the test class S3MinioTest, create the bucket before running the logic.

Regarding recent tickets like https://github.com/linagora/james-project/issues/5234 and the operational habbit to create buckets beforehand, I expect the "create bucket is missing" feature to become mostly irrelevant.

Skipping it in MinIO test (with a nice comment documenting this matter of fact along side the code) seems like a good idea to me.

Thanks a lot for this @vttranlina and option 2. have my +1!

Regards

**Comment by vttranlina** [comment]
> I'm not sure whether this is a bug or a feature of S3 Minio (I’m asking them on their Slack channel and awaiting response).

They answered:
```
Looks like some internal Go stuff was changed... Seems related to https://github.com/golang/go/issues/68199
Connection: close will AFAICT only tell it that it shouldn't reuse the connection. Seems like a bug in the SDK
On the Go side, you are welcome to do your own searching. It may or may not be the linked issue.
```

anw, we don't need to dig deep it. Just simple go (2)

**Comment by Arsnael** [comment]
Thanks for pushing the diagnostic that far @vttranlina . Great catch!

I agree with option 2 too.

**Comment by quantranhong1999** [comment]
> I agree with option 2 too.

+1

**Comment by vttranlina** [comment]
I created a pr for (2) at: https://github.com/apache/james-project/pull/2461

**Pull request closed without merging**